File: packages/sdkwork-assets-core/src/mediaResource.ts
import { isBlank, trim } from '@sdkwork/utils';

export type MediaKind =
  | 'archive'
  | 'audio'
  | 'document'
  | 'image'
  | 'model'
  | 'other'
  | 'video'
  | 'voice';

export type MediaSource =
  | 'data_url'
  | 'drive'
  | 'external_url'
  | 'generated'
  | 'provider_asset';

export type MediaAiProvenanceKind = 'edited' | 'generated' | 'imported' | 'uploaded';

export type MediaModerationStatus =
  | 'approved'
  | 'blocked'
  | 'pending'
  | 'rejected'
  | 'unknown';

export type MediaAccessVisibility =
  | 'organization'
  | 'private'
  | 'public'
  | 'signed'
  | 'tenant';

export interface MediaChecksum {
  algorithm: string;
  value: string;
}

export interface MediaAccess {
  visibility: MediaAccessVisibility;
  expiresAt?: string;
}

export interface MediaAiProvenance {
  provenance?: MediaAiProvenanceKind;
  provider?: string;
  model?: string;
  promptId?: string;
  generationTaskId?: string;
  sourceMediaIds?: string[];
  seed?: string;
  moderationStatus?: MediaModerationStatus;
  safetyLabels?: string[];
}

/** Canonical business-state media per MEDIA_RESOURCE_SPEC §3. */
export interface MediaResource {
  id?: string;
  kind: MediaKind;
  source: MediaSource;
  url?: string;
  publicUrl?: string;
  uri?: string;
  objectBlobId?: string;
  fileName?: string;
  mimeType?: string;
  sizeBytes?: string;
  checksum?: MediaChecksum;
  width?: number;
  height?: number;
  durationSeconds?: number;
  altText?: string;
  title?: string;
  poster?: MediaResource;
  thumbnails?: MediaResource[];
  variants?: MediaResource[];
  access?: MediaAccess;
  ai?: MediaAiProvenance;
  metadata?: Record<string, unknown>;
}

/** Backward-compatible alias used by PC commons consumers. */
export type MediaResourceLike = MediaResource;

export function readMediaResourceUrl(value: unknown): string {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return '';
  }
  const record = value as Record<string, unknown>;
  for (const key of ['publicUrl', 'url', 'uri', 'objectBlobId', 'id']) {
    const raw = record[key];
    if (typeof raw === 'string') {
      const normalized = trim(raw);
      if (!isBlank(normalized)) {
        return normalized;
      }
    }
  }
  return '';
}

export function readMediaResourceThumb(media: MediaResource | undefined): string {
  return readMediaResourceUrl(media?.poster)
    || readMediaResourceUrl(media?.thumbnails?.[0])
    || readMediaResourceUrl(media);
}

export function toExternalUrlMediaResource(
  value: string | null | undefined,
  kind: MediaKind = 'image',
): MediaResource | undefined {
  const url = typeof value === 'string' ? trim(value) : '';
  if (isBlank(url)) {
    return undefined;
  }
  return {
    kind,
    source: url.startsWith('data:') ? 'data_url' : 'external_url',
    url,
    publicUrl: url,
  };
}

export function assertMediaResource(value: MediaResource): void {
  if (isBlank(value.kind)) {
    throw new Error('MediaResource.kind is required');
  }
  if (isBlank(value.source)) {
    throw new Error('MediaResource.source is required');
  }
}

export function normalizeMediaResourceSnapshot(value: unknown): MediaResource | undefined {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return undefined;
  }
  const record = value as Record<string, unknown>;
  const kind = record.kind;
  const source = record.source;
  if (typeof kind !== 'string' || isBlank(kind) || typeof source !== 'string' || isBlank(source)) {
    return undefined;
  }
  return record as MediaResource;
}

File: crates/sdkwork-assets-contract/src/media_resource.rs
use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::media_kind::{MediaKind, MediaSource};

#[derive(Debug, Error, PartialEq, Eq)]
pub enum MediaResourceError {
    #[error("media resource kind is required")]
    MissingKind,
    #[error("media resource source is required")]
    MissingSource,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaChecksum {
    pub algorithm: String,
    pub value: String,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MediaAccessVisibility {
    Private,
    Tenant,
    Organization,
    Public,
    Signed,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaAccess {
    pub visibility: MediaAccessVisibility,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MediaAiProvenanceKind {
    Uploaded,
    Generated,
    Edited,
    Imported,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MediaModerationStatus {
    Unknown,
    Pending,
    Approved,
    Rejected,
    Blocked,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaAiProvenance {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provenance: Option<MediaAiProvenanceKind>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub generation_task_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_media_ids: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub seed: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub moderation_status: Option<MediaModerationStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub safety_labels: Option<Vec<String>>,
}

/// Canonical business-state media descriptor per MEDIA_RESOURCE_SPEC §3.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaResource {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub kind: MediaKind,
    pub source: MediaSource,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub public_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uri: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub object_blob_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub checksum: Option<MediaChecksum>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub height: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_seconds: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alt_text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub poster: Option<Box<MediaResource>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thumbnails: Option<Vec<MediaResource>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variants: Option<Vec<MediaResource>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub access: Option<MediaAccess>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ai: Option<MediaAiProvenance>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

impl MediaResource {
    pub fn validate(&self) -> Result<(), MediaResourceError> {
        let _ = self.kind;
        let _ = self.source;
        Ok(())
    }
}

pub fn media_kind_to_asset_kind(kind: MediaKind) -> &'static str {
    match kind {
        MediaKind::Image => "image",
        MediaKind::Video => "video",
        MediaKind::Audio | MediaKind::Voice => "audio",
        MediaKind::Document => "document",
        MediaKind::Model => "model",
        MediaKind::Archive => "file",
        MediaKind::Other => "other",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn maps_media_kind_to_drive_asset_kind() {
        assert_eq!(media_kind_to_asset_kind(MediaKind::Voice), "audio");
        assert_eq!(media_kind_to_asset_kind(MediaKind::Model), "model");
        assert_eq!(media_kind_to_asset_kind(MediaKind::Archive), "file");
    }
}
